File: frontend/lib/tutor.ts
export interface TutorMessage {
  id: string;
  role: "user" | "assistant";
  content: string;
  metadata: Record<string, unknown> | null;
  created_at: string;
}

export interface Conversation {
  id: string;
  title: string | null;
  skill_id: string | null;
  skill_name: string | null;
  is_active: boolean;
  message_count: number;
  last_message_preview: string | null;
  created_at: string;
  updated_at: string;
}

export interface ConversationDetail extends Conversation {
  messages: TutorMessage[];
}

export interface SendMessageResponse {
  user_message_id: string;
  assistant_message_id: string;
  content: string;
  conversation_id: string;
  metadata: Record<string, unknown>;
}

export interface TutorSkillMastery {
  id: string;
  name: string;
  mastery: number;
  level: "weak" | "developing" | "strong";
}

export interface GoalSkillOption { id: string; name: string; slug: string }

export interface LearnerContext {
  user_name: string;
  goal_title: string;
  preferred_style: ExplanationStyle;
  daily_minutes: number;
  current_phase: string;
  current_focus_skills: string[];
  goal_skills: GoalSkillOption[];
  skill_mastery: TutorSkillMastery[];
  strong_skills: string[];
  weak_skills: string[];
  today_plan_items: string[];
  recent_mistakes: Array<{ exercise: string; skill: string | null; answer: string | null }>;
  total_xp: number;
  streak_days: number;
}

export type ExplanationStyle = "visual" | "mathematical" | "step_by_step" | "analogies" | "balanced";

export interface SuggestedPromptsResponse {
  prompts: string[];
  generated_for_skill: string | null;
}

File: backend/app/schemas/tutor.py
"""Tutor conversation request and response schemas."""

from datetime import datetime

from pydantic import BaseModel, Field, field_validator


class ConversationCreateRequest(BaseModel):
    title: str | None = Field(default=None, max_length=300)
    skill_id: str | None = None


class MessageRequest(BaseModel):
    content: str = Field(min_length=1, max_length=4000)
    socratic_mode: bool = False
    skill_focus: str | None = Field(default=None, max_length=200)
    regenerate: bool = False

    @field_validator("content")
    @classmethod
    def clean_content(cls, value: str) -> str:
        cleaned = value.strip()
        if not cleaned:
            raise ValueError("Message cannot be empty")
        return cleaned


class MessageResponse(BaseModel):
    id: str
    role: str
    content: str
    metadata: dict[str, object] | None
    created_at: datetime


class ConversationResponse(BaseModel):
    id: str
    title: str | None
    skill_id: str | None
    skill_name: str | None
    is_active: bool
    message_count: int
    last_message_preview: str | None
    created_at: datetime
    updated_at: datetime


class ConversationDetailResponse(ConversationResponse):
    messages: list[MessageResponse]


class SendMessageResponse(BaseModel):
    user_message_id: str
    assistant_message_id: str
    content: str
    conversation_id: str
    metadata: dict[str, object]


class SuggestedPromptsResponse(BaseModel):
    prompts: list[str]
    generated_for_skill: str | None


class DeleteConversationResponse(BaseModel):
    deleted: bool
